torch_geometric/nn/models/revgnn.py

```diff
@@ -0,0 +1,288 @@
+            )
+        else:
+            y = fn(*args)
+
+        return self.if_unpack(y)
+
+    @staticmethod
+    def if_unpack(out):
+        # If the layer only has one input, we unpack the tuple again
+        if isinstance(out, tuple) and len(out) == 1:
+            return out[0]
+        return out
+
+
+class GroupAddRev(InvertibleModule):
+    r"""InvertibleModule of Grouped Reversible GNNs for saving GPU memory.
+    Proposed in Training Graph Neural Networks with 1000 Layers:
+    https://arxiv.org/abs/2106.07476.
+    This module enables training arbitary deep GNNs with memory complexity
+    that is independent of the number of layers.
+
+    Args:
+        gnn (Union[nn.Module, nn.ModuleList]):
+            A seed gnn for building gnn groups or a groups of gnns.
+        split_dim (int optional): The dimension to spilt groups.
+        num_groups (Optional[int], optional): The number of groups for
+            group additive reverse.
+        disable (bool, optional): This disables using the InvertibleFunction.
+            Therefore, it executes functions without memory savings.
+        num_bwd_passes (int, optional):
+            Number of backward passes to retain a link with the output.
+            After the last backward pass the output is discarded
+            and memory is freed.
+    """
+    def __init__(
+        self,
+        gnn: Union[nn.Module, nn.ModuleList],
+        split_dim: int = -1,
+        num_groups: Optional[int] = None,
+        disable: bool = False,
+        num_bwd_passes: int = 1,
+    ):
+        super().__init__(disable, num_bwd_passes)
+
+        if isinstance(gnn, nn.ModuleList):
+            n_groups = len(gnn)
+            if num_groups is not None:
+                assert (n_groups == num_groups
+                        ), "Number of GNN groups mismatch with num_group"
+            num_groups = n_groups
+            gnn_groups = gnn
+        else:
+            assert num_groups is not None, "Please specific number of groups"
+            gnn_groups = self._create_groups(gnn, num_groups)
+
+        assert num_groups >= 2, "Number of groups should not be smaller than 2"
+
+        self.gnn_groups = gnn_groups
+        self.split_dim = split_dim
+        self.num_groups = num_groups
+
+    def _forward(self, x: Tensor, edge_index: Adj, *args):
+        xs = torch.chunk(x, self.num_groups, dim=self.split_dim)
+        args_chunks = self._get_arg_chucks(args)
+        y_in = sum(xs[1:])
+        ys = []
+        for i in range(self.num_groups):
+            y_in = xs[i] + self.gnn_groups[i].forward(y_in, edge_index, *
+                                                      args_chunks[i])
+            ys.append(y_in)
+
+        y = torch.cat(ys, dim=self.split_dim)
+
+        return y
+
+    def _inverse(self, y: Tensor, edge_index: Adj, *args):
+        ys = torch.chunk(y, self.num_groups, dim=self.split_dim)
+        args_chunks = self._get_arg_chucks(args)
```

Comment:
This assumes that `args` can be divided into `num_groups`, right? Why do we have this constraint for `args`? I would assume that `args` mostly refer to edge weights or the batch vector.

Reply:
The args will be edge features, dropout_masks, or any tensor that has the same dimension as node features. It can be also absent. Do you mean to chunk the node_features and other args together using `self._get_arg_chucks`. It would be less reabable. But it is ok.

Reply:
This seems to be something I do not understand yet. Currently, it looks like this constraints any other argument to have the same dimension as `x`, which is not true for edge features nor dropout masks, etc. Why do we need to chunk other arguments at all?